{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f278d3a0",
   "metadata": {},
   "source": [
    "<h3> Biological databases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3ef2671",
   "metadata": {},
   "source": [
    "<h4> Exercise1: Write a function that takes PDB ID as argument and saves into the file all lines which start with \"ATOM\" and \"HETATM\".What is the number of CA atoms in \"1X1R\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "30c3f24e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of CA atoms:  169\n"
     ]
    }
   ],
   "source": [
    "from urllib.request import urlopen #module for URL processing\n",
    "\n",
    "def downloadPDB(pdbId):\n",
    "    PDB_URL = f'http://www.rcsb.org/pdb/files/{pdbId}.pdb'   \n",
    "    response = urlopen(PDB_URL)            #open URL \n",
    "    data = response.read().decode('utf-8') #read the content and use utf-8 encoding\n",
    "    return data.split(\"\\n\")\n",
    "\n",
    "def Save_into_file(pdbId):\n",
    "    t = downloadPDB(pdbId)\n",
    "    my_file = open(\"file1.txt\", \"w\")\n",
    "    for elem in t:\n",
    "        if elem.startswith(\"ATOM\") or elem.startswith(\"HETATM\"): \n",
    "            my_file.write(elem)\n",
    "            my_file.write(\"\\n\")\n",
    "    my_file.close()\n",
    "    \n",
    "    count = 0\n",
    "\n",
    "    for elem in t:\n",
    "        if elem.startswith(\"ATOM\"): \n",
    "            if 'CA' in elem:\n",
    "                count += 1\n",
    "\n",
    "    print('Number of CA atoms: ',count)\n",
    "\n",
    "Save_into_file(\"1X1R\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7d57453",
   "metadata": {},
   "source": [
    "<h4> Exercise2: For a given paper (chosen by yourself) save its abstract into a file (from PubMed database)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "70a7db62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Correlation networks are increasingly being used in bioinformatics applications. For example, weighted gene co-expression network analysis is a systems biology method for describing the correlation patterns among genes across microarray samples. Weighted correlation network analysis (WGCNA) can be used for finding clusters (modules) of highly correlated genes, for summarizing such clusters using the module eigengene or an intramodular hub gene, for relating modules to one another and to external sample traits (using eigengene network methodology), and for calculating module membership measures. Correlation networks facilitate network based gene screening methods that can be used to identify candidate biomarkers or therapeutic targets. These methods have been successfully applied in various biological contexts, e.g. cancer, mouse genetics, yeast genetics, and analysis of brain imaging data. While parts of the correlation network methodology have been described in separate publications, there is a need to provide a user-friendly, comprehensive, and consistent software implementation and an accompanying tutorial.\n",
      "The WGCNA R software package is a comprehensive collection of R functions for performing various aspects of weighted correlation network analysis. The package includes functions for network construction, module detection, gene selection, calculations of topological properties, data simulation, visualization, and interfacing with external software. Along with the R package we also present R software tutorials. While the methods development was motivated by gene expression data, the underlying data mining approach can be applied to a variety of different settings.\n",
      "The WGCNA package provides R functions for weighted correlation network analysis, e.g. co-expression network analysis of gene expression data. The R package along with its source code and additional material are freely available at http://www.genetics.ucla.edu/labs/horvath/CoexpressionNetwork/Rpackages/WGCNA.\n"
     ]
    }
   ],
   "source": [
    "from urllib.request import urlopen\n",
    "from lxml import etree #external library for XML files processing\n",
    "\n",
    "PMID = '19114008'\n",
    "\n",
    "baseurl = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "query = f\"db=pubmed&id={PMID}&format=xml\"    \n",
    "url = baseurl+query #concatenation of efetch and our query\n",
    "\n",
    "\n",
    "f = urlopen(url) #open url\n",
    "resultxml = f.read() #read the content\n",
    "\n",
    "xml = etree.XML(resultxml) \n",
    "\n",
    "resultelements= xml.xpath(\"//AbstractText\") \n",
    "\n",
    "f = open(\"Abstract.txt\", \"w\")\n",
    "for element in resultelements:\n",
    "    abstract = ''.join([element.text])\n",
    "    print(abstract)\n",
    "    f.write(abstract)\n",
    "         \n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "046f6a6e",
   "metadata": {},
   "source": [
    "<h4> Exercise3: Find an appropriate tag and download a definition for NM_204410.1 (from  NCBI/nucleotide)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3d0e5ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Gallus gallus secreted protein acidic and cysteine rich (SPARC), mRNA']\n"
     ]
    }
   ],
   "source": [
    "from urllib.request import urlopen\n",
    "from lxml import etree\n",
    "\n",
    "ID = 'NM_204410.1'\n",
    "baseurl= \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "query= f\"db=nucleotide&id={ID}&format=xml\"\n",
    "url= baseurl+query\n",
    "\n",
    "f = urlopen(url) \n",
    "resultxml= f.read() \n",
    "xml= etree.XML(resultxml) \n",
    "\n",
    "resultelements = xml.xpath(\"//GBSeq_definition\") \n",
    "\n",
    "for element in resultelements:\n",
    "    print([element.text])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d0cae28",
   "metadata": {},
   "source": [
    "<h4> Exercise4: Download the record KU325498.1 (from  NCBI/nucleotide) and save it into the file using FASTA format (where ID is a name)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6f7a6bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ID = 'KU325498.1'\n",
    "baseurl= \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "query= f\"db=nucleotide&id={ID}&format=xml\"\n",
    "url= baseurl+query\n",
    "\n",
    "f = urlopen(url) \n",
    "resultxml= f.read() \n",
    "xml= etree.XML(resultxml) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "503c190e",
   "metadata": {},
   "outputs": [],
   "source": [
    "resultelements = xml.xpath(\"//GBSeq_sequence\") \n",
    "\n",
    "f = open(\"record.txt\", \"w\")\n",
    "for element in resultelements:\n",
    "    f.write(f\">{ID}\\n{element.text}\")\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a622495b",
   "metadata": {},
   "source": [
    "<h4> Exercise5: Download SMILES format and plot \"chlordiazepoxide\" structure. What is the application of such a molecule?\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0ab7a777",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAMAAAE5pE7RAAACylBMVEX19fUAAAAAAP9iYmLExMSTk5MxMTE9PT24uLiiovjv7/VSUlKjo6N7e3t6enpmi4tJSfwyzTKWlvj/AADl5eVycnLU1NQQEBAhISG0tLRBQUGDg4Pn5+cODg7v8PC9zMyjo/hSUvzNzc0oKCjq8upp2Gl+nZ2wwsLQ0NCqqqolJSWEhIRLS0teXl5xcXE5OTmXl5fj4+MSEhK8vLzf39+zs7OGhoZZWVksLCzJycmcnJxvb29CQkIWFha/v782NjaIiIhtbW0bGxva2tqKiopcXFwuLi7X19epqalNTU0fHx9ra2uZmZnHx8fIyMiamppsbGwPDw8eHh5MTEyoqKjW1tYtLS1bW1uJiYnm5ubZ2fZiYvvExPeTk/kxMf3h4eEUFBSPj49mZmZ7e/p6evra2vZtbfu/v/cbG/6IiPk2Nv27u7s6OjqCgoJzc3NISEitra3KysorKyuQkJBlZWVWVlafn58dHR3Y2Nj+CwsJCf4bG/1YWPsBAf5UVPvCwvc3N/wHB/53d/rr6/UeHv0ZGf2YmPiOjvkmJv0wMP24uPc9Pf09Pfy4uPj3tbX17e34oKD+AQH6bW313d3+AwP5jo720tL+IyP4r6/8Rkb5jIz7aWkoKP3Nzfdj12Pz9POv5q/x9PFJ0UlF0UVy2nL17+/4lJT5kpL3sbH3r6/9GRn+ExP3s7P+FRX7WFj6dXX7UVH8Ozv8RUX15eX8SUn8Pz/21NT229v+CQn+Bwf8Njb6aGj6YmL8MjL4o6P9Kir5e3v4lpb4mpr5fX39MDD5hYX5gYH22dn6bGz22NjE68SU4ZST4ZPF68Vi12LL7Mux57HI68i86bxE0ES26LbG68Y1zTV12nWH3oeG3oZ02nQzzTPA6sB423h73Hte1l47zjvi8eLk8eQ4zjg+zz7l8eXf8N82zTZc1VyD3YOL34ul5KXa79rQ7dBG1HNLAAAACXBIWXMAAA7EAAAOxAGVKw4bAAAN3ElEQVR4nO2dh3/UOBaA5UkmMySZBAgtEPYC7O2G0AkhEAIhhL579AWOvkvvfYG723J7d3t7wN723sv13nvvvffe+/9wI3uK7ZE0kixbmvH7fuAZW7bm5VmWVd57QqgqsTjOacr+b1T1g2nn40SZ0+ao+j2bWzjPs/C/TVzn8ZBmpjbfz5dLEF4aRjh4l+v7WN6calFdcHFU0SB11QJawih7O8XeJun3djpKoducrzVSAhRJKD0N7ZQWBIgF7JrIQ7maPhNIECKb1WW1tPwp65mpnyyfQTNCD/FJ85S93ch3cjnqh2Q3/Zwnz0ZJNb9altEqM8sJ3R+0fsXMQKg9X4PWBsuKqx6+O9hveDHoZQsAflJIthHmoSl4FkR4a0kKW+apEaOUU4Gu9r7FXx0oL4T2BLzei91sXhVUphwpucua8aYFPahChGbCsXVBM70SNAM39XfgBsHJW32Hve/9XyNSD5DNZOfDfgXVFPq63+W9vN27O8W9041qRwpIoqCKyaO0mVIhTEOunupC/r5+jg5CrwFnJ96Ymlb8ytulDZ3EEN0SVB5R9SkAAAAIrFWV0bKbFGW0WlE+XiYGvD7jfnkH7Gc1edrjk4JlFhpng1y8wrt7Pkheq3uCXO3lRnVZhcgqZTn1oBvUZJTpRmi8mqxQm6IRXLRyq6KMAvKhz6vJx+4kf8lzSLLrjvNqRh9t8RwSmLXw51UptCnMazk96TnOLBLDs5ujZ5Cv6z3OvfMjzryG3lP4WswMDyoEG4wfcO/IZLXf9d0ZKMx3ayyntzxKRixkz5y5h+17UI1If0lkXCRC4jARYezfGGaZmOF+dnq9j3hZWl3f7QfSlVmAx3tC7tMZc5QZbSC9frBAkhOK6SP+EataqZxm5T4TwYe/pgbOIc+udlU5bVM4qne7yhFCY0Ybw2SXbgFowHAhAAAAAACAoWQ2pI0a11+R2uLaC8HiWBiLNBAfdI4mJAYVuutIYVFGgQeUWvRIQBtQtrhdetRDU5aDyvFhIdJoLj3xQnRyeMmgbkbq6cjk8JFiFZ9j0cnhYzcrUedU1M3UlJURSkGA9iiqmkmUxle+5s6zHSRfpUMUL45gGcsxMVi32QCRHGzB1ixXNJ2sDulJzXCRnrcNxrvvQ1/4Bj05WmU1F2aOWxD6yiP0E6NVVjPnhPaasAVh8jgtge1vFz6PkQ9zuHmHzL2EYxz+iyR+9sd/Shj6U6h/DRruMo6+eG78cbkOxQ9/md2oE2toAot1+eoVt+IoDa86hGqoMv/mz//+r0KxCKbkjYOlGks60yFY4DEjSrJRJk8eytSX5bHG98x819T6pjdD8Jk4Qk0ZsD02/TI5zMEOWfadbmgtTQ3OAVZitjvdQZ/JYvSSgmLm9FkoNyAwsjY78cRUbZlZtkx9Eo21GWOM8g02og7sX0h8bc/vy745Vbj+06AYelk4XkKeGq/9UvaSEWNCFMmG9KJtLzlSeDfXlUgZClNQJ0qnUoU+WGIb1Z7GKmzCZjbqcu3tP5BAiduH6DbzyZbomf5jCXS3ZrF26hvUZtKuWwAfiduMtBNLqLAWjBWmGsJBeC4hTG1tmWombqpcAAAAAAAAAAAAAAAAQEXRtoppT6wFx77TLJPKlevzptV0M+eoWTq+YO5m8S+HEi4+W2FLmw2qm40lEYwt1ZGSJSB5QBkgF3F6Uf99bCIK0K+/3BPrqQG99USGVns2aq1X6V5Rg2X8WkKFXLRyHI9OjhIYt0qb78paNJ+Rei4yOXyk6AtBZbkYmRw+NqCF9ER/kPfoaEJ99MQz0cnh4ybUS088GZ0cXtgV5tGIpChhGStRn9cI0zop8FoL8rAc4XU6szB8p1UFKJWC7p+v18eG5v+hzac6ByFsQLcl7foTGnPm2b4NjKii0ZNO2bGVvfEh9JK3kE2vDms1I3nWEsNW6MWONWIYy1KqgsGrxIDhEBLm3T6MLqnee+nTrGRN9/Bt70PoYUZ6pGJdcx26/hX2t7d/wHatphLpTSyK9c4PMsWKtmjlxGrOCvTAJz77dfqJ0RatorbKoHAxVw58YtHvot7I+0/QEhQtmCfLk+TDmqOwIPQo8ai6ZSYkuUw8qj04DLpKOigZ+eTZvz5/7SuDCFMGuQAxL76AkDqxfJ5st545eVQu+h322lcoVtER6fj4c85IJL1zzRg+USyW42C3N7HPdXAiOfhdTS3DXfKnv//7ywrFIvhHLSaudZRznKwVXQBRCqLf3xLCqFLu/iWJPq/fUy8YiQaEOj0Hin6l2KtknM8V59koZMIsGkTTi3vemPt1yOvykgzBC/oQLaERdeVj+fufvyQeLskL1hpKoMOD1JSBnAcqyS+4pzsXfGRUOJ5CExhp/XjBAtqaCTiIWV1oK7MeZiWyIz6GGA+y7Dob7B8OS6yyXvdaxCofdUiLWOXfIzrE4ogIoEMsjuAJGsQqDcAk+sOhiMWTKWirgJlly9Qn0dB6y9Ra3tB3oqktiLJyaRKLdRcX4yA31EYxtscJL4IKvdQvccIRJYlxEPrmE8MHKoNaRzQUAgeWKmxBb/bCcCMSkXN/ZhGa3FHY8wqGex30GJCKIL6B9uZDGeZw38lsgUqKrbUnBaGEPF0aEzKvMNxrjCKKRKdFGs4iRLmp832GynSEOlBTUTRf/FM39p0kP5iq6cJ1k02TdQCvWVqfuIMmFlZURFFAZnmKdvuRfFhWvcx0BXBzqB961z26xZrqDrXqQIwWGy0WDldmHNt1C0BkB3t0Sx+mScWoM3XCrDP1YUSdWYoJdSYBA+rMCsPQuIE7dAtAwcx3j6limXobzQx2GmQ59FAJM85yAAwVy9TbCIVeCFNLFwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABULyuWpsZv2YT2kEPOADnS61ZvbiqGV0qF7ftWmbStXWVtvLE07tot5oVv1cmazPINN5AjH9o2hzeHEuCm4li5bH1qKzt4rK2utFx8sWoizRew0lbXbsu8taYiJcUbdNQxAJ5kZlDeaLALFqch9JJFeNtgzspcUYMLVob7VdewGG8zuhdO0oRIwbJZbL8UGw0N3B0uYgXLZtESvJ24JwxxjEa4YBU5ekytKOYjUbDynDylWBbTWYOjb8sVrNP6FnPSxLqtCPX2SF16Xt+CXJpYn+0x93XLXHnxgmpZTGflRoTmMFZzY3DurGJZjGfVJoTmslbko9KmP/B6xNh/8TzWqopUTuhcnVILSyUD4CNjvWpDZLlcWP4sa81ZiisypGciuId1qgq5mYgVG1TLUSHIzERIdpCqAfGZiNUmrvYWFYIzEZyj9lXL7rJtgfkLFvblFiNeZt5yi1FDV1dv38IFhaYrnoWNRB7D8SshnUlZqaZCK4ExCxtLCupyK4lrFjaWeEtXzowGoOGoy2dGA9AgmtEARGLcPhcHWgb8xKtgveuBD3/8oc+0fE3y8qotWNc0X5fdXt/sXqH3rS3vxx9fbHlEKsvqLVgkZb2j5UH88eWWb0plWbUFK6ssB7ey7nvPpY9kP7768LdkcqzegkUsWdkH8WOXPvW5++VyrN6CRVFWAOI4E+Hw6GPCl8RzJsLhyafEzo/tTITDE4+LnF3F1TsfIsH3Yz0TYbN3X/lzHOI+E2HDW7ginYn48c9/96e//WvYf64dpmxx98hoO3XimBWhY+EPhv3C+WKesuoTidchNJyw1srps+cvXDh/9rSz1xBZDf/8sF85X0xU1msTd3qUdXnvVevYiVNt/jMbBV2/autq0MjR4mvMvfCTl56zv5iorKHbXv/GNzvKSlzZdy/5NMt2zpnEbS2StPJaGtVQ1yoo02//8Jd/vPjy/4xUFnrDmxLllzwasA0fMjzmDzV1vhUZWusaRklJZxhYWQi9hWd9qMEluGBNtDpYJ+GHz0Nec8nRI8v9wHe+X16ICmJxP95OnUJJLj585MMlRc7FOOvbwWSLgsMHxM63XQtnzihNYGmiWOCI+mwcUxPiWr/qaBdfj9B2LZw1u8t1qOThI7FgYS/h5NqxIa8/rI6DR2SuclwLp0139igPH4n5fX2O1Y1TDEfUjKmguBD7D0le6LgWTu5kP3xEehc63ghJK4J1mlUSeKw4SAYVNlA9OnC1Gh9ljQneUoyNsoRrGwJxUdZY0S4biZgoq0HJSzseyqoRHzshEQtljVTUxgFlCRALZcFjKARU8CJA00EEaJSKAN0dEaAjLQIM0QgwYrTslYvsP7WLf5DUxVxHTZWmLDRqjNRlDQN4OxWPK3MNvxfpntczB39yzIyZR22d8CWNSwaz25mdswpHOMeW5/TMswMnVe6ca+tYsfP7F2c3szpn+g6XnbVIWnPxh8Rsvkk0CgQkaH8ab6dPJSYyHshc0ZOxEzGMJG+lM8G2FLS6GKcQHsicDgXrNmOxuOraxDP2tux53gfS2RGYWjSbqZMRSlspK5Upb+rOaVbpLUXiU4umMnMacvRlk24qURrdNJCNo6Jqefgws2ajnL68HDw84dChCYcPIpJpIDdV8/DZzOjK6YuBxzQwvkzBDfAZrHcbhts0sJrp6EQ5fbERMA2sWnbtyOkLKMfO7dnNhHbNUlQGicIG4EHCPDJWyDYvY0mA5mX8gOalANC8FACal0C4bN+pW4JKAnd1AF6g8S7Ajl26Jagk7I40wIkiK8mYUDUzLpFQcTYsOpGwCIkx3LP2AIJKXgyo5EWASh4AAAAAAAAAAAAAAAAAAACIN/8HKKk77giEtQwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<IPython.core.display.Image object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from urllib.request import urlretrieve\n",
    "from IPython.display import Image\n",
    "\n",
    "molecule = \"chlordiazepoxide\"\n",
    "query = f\"https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/name/{molecule}/XML\"\n",
    "\n",
    "response = urlopen(query)\n",
    "resultxml = response.read()\n",
    "tree = etree.XML(resultxml)\n",
    "\n",
    "sek = []\n",
    "for element in tree.iter():\n",
    "    if element.tag == \"{http://www.ncbi.nlm.nih.gov}PC-InfoData_value_sval\": #there is a SMILES format\n",
    "        sek.append(element.text)\n",
    "\n",
    "smiles = sek[-2] #aspirin molecule in SMILES format\n",
    "urlretrieve('https://pubchem.ncbi.nlm.nih.gov/rest/pug/compound/smiles/'+smiles+'/PNG', 'smi_pic.png')\n",
    "p = Image(filename='smi_pic.png')\n",
    "p"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96a2a66e",
   "metadata": {},
   "source": [
    "$\\to$ Chlordiazepoxide is indicated for the short-term (2â€“4 weeks) treatment of anxiety that is severe and disabling or subjecting the person to unacceptable distress. It is also indicated as a treatment for the management of acute alcohol withdrawal syndrome."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb350690",
   "metadata": {},
   "source": [
    "<h4> Exercise6: Explain the following patterns and give example strings that match them:</h4>\n",
    "    \n",
    "        \n",
    "    - AT*C:     ATC, ATAAC \n",
    "    - AT+C:     ATGC, ATAAC\n",
    "    - AT?C:     ATC, ATAC\n",
    "    - AT{3}C:   ATTTC\n",
    "    - AT{2,8}C: ATTC, ATTTTTC\n",
    "    - A.*C:     AC, ATGC, AAAAAC\n",
    "    - A.+C:     ATC, ATATC\n",
    "    - A.?C:     AC, AGC\n",
    "    - A.{3}C:   ACCCC, ATCTC\n",
    "    - A.{2,8}C: ATTC, ATGTAAC\n",
    "    - A[GC]T:   AGT, ACT\n",
    "    - A[^G]T:   ACT, ATT, AAT\n",
    "    - ^A[GC]T$: AGT, ACT\n",
    "    - ^A[^\\s]+: ATGCA, AA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "890b8345",
   "metadata": {},
   "source": [
    "<h4> Exercise7: Write a regular expression that encompasses the following words:</h4>\n",
    "        <br>\n",
    "        \n",
    "    - ATA, AGA, ACA:           A.A\n",
    "    - AT, ACT, ACCT, ACCCCT:   A[CT]+\n",
    "    - TAC, TCC, TCAAC, TCAACC: T[AC]{2,5}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "529506eb",
   "metadata": {},
   "source": [
    "<h4> Exercise8: Using regex find all non-overlapping subsequences composed of only C residues. For example, after application to ACTCCGCCCC you should obtain ['C','CC','CCC']."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "112e8935",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['C', 'CC', 'CCCC']\n"
     ]
    }
   ],
   "source": [
    "import regex as re #library for regular expressions\n",
    "\n",
    "seq = \"ACTCCGCCCC\"\n",
    "pattern = 'C+'\n",
    "print(re.findall(pattern, seq)) #findall has two arguments --> pattern and text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a65c4350",
   "metadata": {},
   "source": [
    "<h4> Exercise9: Using regex, remove all low characters from a sequence. For example, after application to ATTAAAaaaGccT you should obtain ATTAAAT. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "db50507f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTAAAGT\n"
     ]
    }
   ],
   "source": [
    "seq = \"ATTAAAaaaGccT\"\n",
    "pattern = '[^[A-Z]]*'\n",
    "print(re.sub(pattern, '', seq)) #sub has three arugments: pattern, to what we change the pattern, and text"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cbe6b07",
   "metadata": {},
   "source": [
    "<h4> Exercise10: Consider the motif| A or G | ? | ? | ? | ? | G | K | S or T |.  \n",
    "    \n",
    "- Write a regular expression for the motif. $\\to$ [AG].{4}GKS[ST]\n",
    "- Download sequences AAH10041.2 and CAG38727.1 (from NCBI/proteins, automatically).\n",
    "- What are the names of these proteins? \n",
    "- Localize the motif fragments within these proteins (positions).\n",
    "- Compare the results with UNIPROT database data (Features)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6c94f824",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Download sequences AAH10041.2 and CAG38727.1 (from NCBI/proteins, automatically).\n",
    "\n",
    "ID1 = 'AAH10041.2'\n",
    "ID2 = 'CAG38727.1'\n",
    "\n",
    "def download(ID, name):\n",
    "    \n",
    "    baseurl= \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "    query= f\"db=Protein&id={ID}&format=xml\"\n",
    "    url= baseurl+query\n",
    "\n",
    "    f = urlopen(url) \n",
    "    resultxml= f.read() \n",
    "    xml= etree.XML(resultxml) \n",
    "        \n",
    "    resultelements = xml.xpath(\"//GBSeq_sequence\") \n",
    "\n",
    "    f = open(name, \"w\")\n",
    "    for element in resultelements:\n",
    "        f.write(f\">{ID}\\n{element.text}\")\n",
    "    f.close()\n",
    "    \n",
    "    for element in resultelements:\n",
    "        return(element.text)\n",
    "    \n",
    "seq1 = download(ID1, 'ID1_seq')\n",
    "\n",
    "seq2 = download(ID2, 'ID2_seq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a508feb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GBReference_authors', 'GBSeq_moltype', 'GBInterval_accession', 'GBSeq_organism', 'GBXref', 'GBXref_dbname', 'GBReference', 'GBSeq_update-date', 'GBAuthor', 'GBSeq_length', 'GBXref_id', 'GBFeature_key', 'GBQualifier', 'GBSeq_comment', 'GBInterval_from', 'GBSeq_definition', 'GBFeature_location', 'GBQualifier_value', 'GBQualifier_name', 'GBSeq_feature-table', 'GBSeq_taxonomy', 'GBSet', 'GBSeq_other-seqids', 'GBReference_consortium', 'GBReference_title', 'GBFeature_quals', 'GBReference_journal', 'GBSeq_topology', 'GBReference_position', 'GBReference_remark', 'GBFeature_intervals', 'GBSeq_accession-version', 'GBInterval_to', 'GBSeq_locus', 'GBKeyword', 'GBReference_pubmed', 'GBFeature', 'GBSeq_keywords', 'GBReference_xref', 'GBInterval', 'GBSeq_references', 'GBSeq_sequence', 'GBSeq_create-date', 'GBSeq_source-db', 'GBReference_reference', 'GBSeq_source', 'GBSeqid', 'GBSeq_primary-accession', 'GBSeq_division', 'GBSeq']\n"
     ]
    }
   ],
   "source": [
    "tags = list({elem.tag for elem in xml.iter()})\n",
    "print(tags)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f84fea74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Tu translation elongation factor, mitochondrial [Homo sapiens]']\n",
      "['RAB1A [Homo sapiens]']\n"
     ]
    }
   ],
   "source": [
    "# What are the names of these proteins? \n",
    "\n",
    "def give_names(ID):\n",
    "    \n",
    "    baseurl= \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?\"\n",
    "    query= f\"db=Protein&id={ID}&format=xml\"\n",
    "    url= baseurl+query\n",
    "\n",
    "    f = urlopen(url) \n",
    "    resultxml= f.read() \n",
    "    xml= etree.XML(resultxml)\n",
    "    \n",
    "    resultelements= xml.xpath(\"//GBSeq_definition\") \n",
    "\n",
    "    for element in resultelements:\n",
    "        print([element.text])\n",
    "        \n",
    "give_names(ID1)\n",
    "give_names(ID2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "7902e143",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ghvdhgkt']\n"
     ]
    }
   ],
   "source": [
    "# Localize the motif fragments within these proteins (positions).\n",
    "\n",
    "f = open(\"ID1_seq\")\n",
    "t = f.read()\n",
    "f.close()\n",
    "\n",
    "seqs = re.findall(\"[ag].{4}gk[st]\", t) \n",
    "print(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "67b1b293",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pozycja fragmentu ghvdhgkt w sekwencji to 66\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "                  \n",
    "fragment = \"ghvdhgkt\"\n",
    "position = seq1.find(fragment)\n",
    "\n",
    "print(\"Pozycja fragmentu {} w sekwencji to {}\".format(fragment, position))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4b5004bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['gdsgvgks']\n"
     ]
    }
   ],
   "source": [
    "f = open(\"ID2_seq\")\n",
    "t = f.read()\n",
    "f.close()\n",
    "\n",
    "seqs = re.findall(\"[ag].{4}gk[st]\", t) \n",
    "print(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "fe89f049",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pozycja fragmentu gdsgvgks w sekwencji to 17\n"
     ]
    }
   ],
   "source": [
    "fragment = \"gdsgvgks\"\n",
    "position = seq2.find(fragment)\n",
    "\n",
    "print(\"Pozycja fragmentu {} w sekwencji to {}\".format(fragment, position))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "161775ca",
   "metadata": {},
   "source": [
    "- Compare the results with UNIPROT database data (Features).\n",
    "\n",
    "$\\to$ after checking both records in the UNIPROT database, the results obtained in the previous part of the task were confirmed"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
